protocols/rendezvous/src/client.rs

```diff
@@ -400,6 +406,27 @@ impl Behaviour {
             DiscoverResponse(Ok((registrations, cookie))) => {
                 if let Some((rendezvous_node, _ns)) = self.waiting_for_discovery.remove(request_id)
                 {
+                    self.events
+                        .extend(registrations.iter().flat_map(|registration| {
+                            let peer_id = registration.record.peer_id();
+                            registration
+                                .record
+                                .addresses()
+                                .iter()
+                                .filter(|addr| {
+                                    !self.discovered_peers.iter().any(
+                                        |((discovered_peer_id, _), addrs)| {
+                                            *discovered_peer_id == peer_id && addrs.contains(addr)
+                                        },
+                                    )
+                                })
+                                .map(|address| ToSwarm::NewExternalAddrOfPeer {
+                                    peer_id,
+                                    address: address.clone(),
+                                })
+                                .collect::<Vec<_>>()
```

Comment:
```suggestion
                                .map(|address| ToSwarm::NewExternalAddrOfPeer {
                                    peer_id,
                                    address: address.clone(),
                                })
```
I think we don't need to collect here?

Reply:
If I remember correctly, without collecting on the iterator, it would give a borrow error with `peer_id` being used like it does. Could probably do something like the following to make it cleaner. 

```diff
diff --git a/protocols/rendezvous/src/client.rs b/protocols/rendezvous/src/client.rs
index 086491bd2..76bb2e338 100644
--- a/protocols/rendezvous/src/client.rs
+++ b/protocols/rendezvous/src/client.rs
@@ -413,18 +413,18 @@ impl Behaviour {
                                 .record
                                 .addresses()
                                 .iter()
-                                .filter(|addr| {
+                                .scan(peer_id, |peer_id, e| Some((*peer_id, e.clone())))
+                                .filter(|(peer_id, addr)| {
                                     !self.discovered_peers.iter().any(
                                         |((discovered_peer_id, _), addrs)| {
-                                            *discovered_peer_id == peer_id && addrs.contains(addr)
+                                            discovered_peer_id == peer_id && addrs.contains(addr)
                                         },
                                     )
                                 })
-                                .map(|address| ToSwarm::NewExternalAddrOfPeer {
+                                .map(|(peer_id, address)| ToSwarm::NewExternalAddrOfPeer {
                                     peer_id,
-                                    address: address.clone(),
+                                    address,
                                 })
-                                .collect::<Vec<_>>()
                         }));
 
                     self.discovered_peers
```

Thoughts?

Reply:
The use of `scan` confuses me a bit. Can't you just deref / copy `peer_id` in `NewExternalAddrOfPeer` if that's the problem? 

---

Looking through it again, it might still conflict due to `self.discovered_peers` being borrowed multiple here immutably and below mutably. So I think the collect is needed after all.